scanner/updater/rhel/parser.go

```diff
@@ -0,0 +1,241 @@
+				NormalizedSeverity: common.NormalizeSeverity(def.Advisory.Severity),
+				Repo: &claircore.Repository{
+					Name: affected,
+					CPE:  wfn,
+					Key:  repositoryKey,
+				},
+				Dist: u.dist,
+			}
+			vs = append(vs, v)
+
+			// If this is an unfixed OpenShift 4.x vulnerability, add a CPE for each minor version
+			// below the given minor version.
+			// There is only a single OVAL v2 file for all OpenShift 4 versions for each RHEL version,
+			// and it is assumed the CPE specified for the vulnerability indicates
+			// versions y such that 4.0 <= y <= 4.x are affected, where x is the next,
+			// unreleased minor version of OpenShift 4 specified in the CPE.
+			//
+			// It is expected the CPE is of the form cpe:/a:redhat:openshift:4.x or
+			// cpe:/a:redhat:openshift:4.x::el<RHEL version>.
+			// For example: cpe:/a:redhat:openshift:4.14 or cpe:/a:redhat:openshift:4.15::el9.
+			//
+			// Any other OpenShift 4-related CPEs are not supported at this time.
+			if defType == ovalutil.CVEDefinition && strings.HasPrefix(affected, "cpe:/a:redhat:openshift:4") {
+				if openshiftCPEs, err := allKnownOpenShift4CPEs(affected); err != nil {
+					zlog.Warn(ctx).Msgf("Skipping addition of extra OpenShift 4 CPEs for the unpatched vulnerability %q: %v", def.Title, err)
+				} else {
+					for _, openshiftCPE := range openshiftCPEs {
+						wfn, err := cpe.Unbind(openshiftCPE)
+						if err != nil {
+							return nil, err
+						}
+						v := *v
+						v.Repo = &claircore.Repository{
+							Name: openshiftCPE,
+							CPE:  wfn,
+							Key:  repositoryKey,
+						}
+						vs = append(vs, &v)
+					}
+				}
+			}
+		}
+		return vs, nil
+	}
+	vulns, err := ovalutil.RPMDefsToVulns(ctx, &root, protoVulns)
+	if err != nil {
+		return nil, err
+	}
+	return vulns, nil
+}
+
+// severity returns the urlencoded Severity, CVSS scores and vectors based on the
+// OVAL definition. For advisories, the CVSS score is the maximum of all the
+// related CVEs' scores.
+func severity(ctx context.Context, def oval.Definition) string {
+	var cvss3, cvss2 struct {
+		score  float64
+		vector string
+	}
+
+	// For CVEs, there will only be 1 item in this slice.
+	// For RHSAs, RHBAs, etc., there will typically be 1 or more.
+	for _, cve := range def.Advisory.Cves {
+		if cve.Cvss3 != "" {
+			score, vector, found := strings.Cut(cve.Cvss3, "/")
+			if !found {
+				zlog.Warn(ctx).
+					Str("CVSS3", cve.Cvss3).
+					Msg("unexpected format")
+				continue
+			}
+			parsedScore, err := strconv.ParseFloat(score, 64)
+			if err != nil {
+				zlog.Warn(ctx).
+					Str("Vulnerability", def.Title).
+					Err(err).
+					Msg("parsing CVSS3")
+				continue
+			}
+			if parsedScore > cvss3.score {
+				cvss3.score = parsedScore
+				cvss3.vector = vector
+			}
+		}
+
+		if cve.Cvss2 != "" {
+			score, vector, found := strings.Cut(cve.Cvss2, "/")
+			if !found {
+				zlog.Warn(ctx).
+					Str("CVSS2", cve.Cvss2).
+					Msg("unexpected format")
```

Comment:
can you add the vuln title to this log and the log we have for CVSS3, too?

Reply:
34502f8bc8